test/onnx/test_pytorch_onnx_shape_inference.py

```diff
@@ -284,5 +286,171 @@ def test_reduce_prod_without_axes(self):
+            self.assertTrue(dims[i].HasField("dim_param"))
+
+    def test_setType_maintains_output_shape_for_single_custom_op_with_dynamic_axes(
+        self,
+    ):
+
+        self.addCleanup(torch.onnx.unregister_custom_op_symbolic, "::linalg_inv", 9)
+
+        class CustomInverse(torch.nn.Module):
+            def forward(self, x):
+                return torch.inverse(x) + x
+
+        def linalg_inv_settype(g, self):
+            return g.op("com.microsoft::Inverse", self).setType(
+                self.type().with_dtype(torch.float).with_sizes([None, 3, 3])
+            )
+
+        torch.onnx.register_custom_op_symbolic("::linalg_inv", linalg_inv_settype, 9)
+        model = CustomInverse()
+        x = torch.randn(2, 3, 3)
+        f = io.BytesIO()
+        torch.onnx.export(
+            model,
+            (x,),
+            f,
+            opset_version=self.opset_version,
+            custom_opsets={"com.microsoft": 1},
+            input_names=["x"],
+            dynamic_axes={"x": {0: "batch"}},
+        )
+
+        model_proto = onnx.load(io.BytesIO(f.getvalue()))
+        model_value_info = model_proto.graph.value_info
+        self.assertIsNotNone(model_value_info)
+        assert model_value_info
+        dims = model_value_info[0].type.tensor_type.shape.dim
+        # The first axe should be dynamic as we defined when exporting
+        self.assertTrue(dims[0].HasField("dim_param"))
+        for i in range(1, len(dims)):
+            # If node output has shape info, it should have dim_value
+            # Otherwise, it has dim_params with dynamic shape
+            self.assertTrue(dims[i].HasField("dim_value"))
+            self.assertEqual(dims[i].dim_value, x.size()[i])
+
+    def test_setType_maintains_output_shape_for_single_custom_op_with_onnx_ops(self):
+
+        self.addCleanup(torch.onnx.unregister_custom_op_symbolic, "::linalg_inv", 9)
+
+        class CustomInverse(torch.nn.Module):
+            def forward(self, x, y, z):
+                x = torch.inverse(x)
+                return x + y + z
+
+        def linalg_inv_settype(g, self):
+            return g.op("com.microsoft::Inverse", self).setType(
+                self.type().with_dtype(torch.float).with_sizes([2, 3, 10, 10])
+            )
+
+        torch.onnx.register_custom_op_symbolic("::linalg_inv", linalg_inv_settype, 9)
+        model = CustomInverse()
+        x = torch.randn(2, 3, 10, 10)
+        y = torch.randn(2, 3, 10, 10)
+        z = torch.randn(2, 3, 10, 10)
+        f = io.BytesIO()
+        torch.onnx.export(
+            model,
+            (x, y, z),
+            f,
+            opset_version=self.opset_version,
+            custom_opsets={"com.microsoft": 1},
+        )
+
+        model_proto = onnx.load(io.BytesIO(f.getvalue()))
+        # To validate the shape of inverse Op, we need to find inverse output name,
+        # and then use it to identify its value_info for the shape.
+        output_name = ""
```

Comment:
Should we assert on `output_name` not empty str after the loop? It's a bit paranoid but this test will silently pass if `Inverse` is not found in exported graph, as the rest of asserts are mostly skipped from not entering `if value_info.name == output_name:`

Reply:
Added assert output_name